### PR TITLE
pkg/kubelet/cm: cgroup-related cleanups

### PR DESCRIPTION
**Pull request by kolyshkin**
#### What type of PR is this?

/kind cleanup

#### What this PR does / why we need it:

While working on #102147, I noticed a few odd things in the code that can be straightened out.
This PR does some of that. Please see detailed description in commits.

#### Which issue(s) this PR fixes:

none

#### Special notes for your reviewer:

Please review on commit by commit basis.

#### Does this PR introduce a user-facing change?
```release-note
NONE
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

none

### CODE REVIEW

File: pkg/kubelet/cm/cgroup_manager_linux.go
Comment (fromanirh) [inline pkg/kubelet/cm/cgroup_manager_linux.go:378]: this intuitively makes sense to me, but could you please add a reference to the specs/docs (a line in the git commit is fine like `see: https://....`) for future reference?

File: pkg/kubelet/cm/cgroup_manager_linux.go
Comment (odinuge) [inline pkg/kubelet/cm/cgroup_manager_linux.go:443]: Have you tested this with systemd? I thought this was required for that, but maybe that has changed?

File: pkg/kubelet/cm/cgroup_manager_linux.go
Comment (odinuge) [inline pkg/kubelet/cm/cgroup_manager_linux.go:600]: Is this correct? Shouldn't this be `memory.current`?

**Review by odinuge — CHANGES_REQUESTED** [review]
Overall change is ok, but still has some potential to break stuff. This is not urgent, so I think we should wait for https://github.com/kubernetes/kubernetes/issues/102148 and possibly https://github.com/kubernetes/test-infra/pull/22290.

This should also be tested properly on systemd and on cg v2.

/hold

File: pkg/kubelet/cm/cgroup_manager_linux.go
Comment (odinuge) [inline pkg/kubelet/cm/cgroup_manager_linux.go:596]: Do you know of a test signal that can verify this change works as expected?

File: pkg/kubelet/cm/types.go
Comment (odinuge) [inline pkg/kubelet/cm/types.go:81]: Maybe specify that it is in bytes and not pages/kB/MB?

File: pkg/kubelet/cm/cgroup_manager_linux.go
Comment (kolyshkin) [reply pkg/kubelet/cm/cgroup_manager_linux.go:378]: I was going to refer to https://github.com/opencontainers/runc/pull/2490, but unfortunately it falls short in case of systemd -- one more PR is needed (created: https://github.com/opencontainers/runc/pull/2958).

Removing this commit until https://github.com/opencontainers/runc/pull/2958 is merged, a release is cut, and kubernetes vendoring is updated to include it.

File: pkg/kubelet/cm/cgroup_manager_linux.go
Comment (kolyshkin) [reply pkg/kubelet/cm/cgroup_manager_linux.go:443]: As explained above (https://github.com/kubernetes/kubernetes/pull/102218#discussion_r637454215), I rushed to make this change. Removing this commit until https://github.com/opencontainers/runc/pull/2958 is merged, updated runc is released and vendored in here.

File: pkg/kubelet/cm/cgroup_manager_linux.go
Comment (kolyshkin) [reply pkg/kubelet/cm/cgroup_manager_linux.go]: 🤦🏻 my bad; fixed

File: pkg/kubelet/cm/cgroup_manager_linux.go
Comment (kolyshkin) [reply pkg/kubelet/cm/cgroup_manager_linux.go:606]: The code that is using this was added by #41149, and later moved out of experimental in #62509. Neither comes with any tests, but maybe @sjenning can shed more light on it (he should review this anyway).

File: pkg/kubelet/cm/types.go
Comment (kolyshkin) [reply pkg/kubelet/cm/types.go:81]: Done. I have also removed that the value is read from cgroupfs as this is kind of obvious.

**Comment by kolyshkin** [comment]
```
E2eNode Suite: [sig-node] Summary API [NodeConformance] when querying /stats/summary should report resource usage through the stats api expand_less	1m55s
_output/local/go/src/k8s.io/kubernetes/test/e2e_node/summary_test.go:54
Timed out after 90.000s.
Expected
    <string>: Summary
to match fields: {
[.Node.SystemContainers[runtime].Memory:
	Expected
	    <string>: MemoryStats
	to match fields: {
	.AvailableBytes:
		Expected
		    <*uint64 | 0xc000a91418>: 3705806848
		to be nil
	}
	, .Node.SystemContainers[pods].Memory:
	Expected
	    <string>: MemoryStats
	to match fields: {
	.MajorPageFaults:
		Expected
		    <uint64>: 198
		to be <=
		    <int>: 10
	}
	, .Node.SystemContainers[kubelet].Memory:
	Expected
	    <string>: MemoryStats
	to match fields: {
	.AvailableBytes:
		Expected
		    <*uint64 | 0xc000a91560>: 3836538880
		to be nil
	}
	]
}
```

As far as I understand, there are two issues here.

1. The test expects `MemoryStats.AvailableBytes` to be `nil` for system containers, and yet it reports something close to 4GB for `runtime` and `kubelet` (but not for `pods`).
2. Higher than expected number of `MemoryStats.MajorPageFaults` from the `pods` system container.

This is _presumably_ caused by the first commit (removing `propagateControllers`), and I can't think about why is that. Perhaps somehow writing to `cgroup.subtree_control` reset the state of the controller?

Looking at it.

**Comment by odinuge** [comment]
> As far as I understand, there are two issues here.

Ahh, sorry, thought you knew about these test failures... They are expected, I should probably have stated that explicitly. The failures in `pull-kubernetes-node-crio-cgrpv2-e2e ` will be fixed by bumping cadvisor and the commit changing the conformance test in your PR. They are tracked here: https://github.com/kubernetes/kubernetes/issues/99230

`pull-kubernetes-node-kubelet-serial-crio-cgroupv{1,2}` are just experimental, so don't care about those... :sweat_smile: 


Overall I think this PR is ok. There are some issues with controller propagation on v2 in general (some WIP stuff here: https://github.com/kubernetes/kubernetes/pull/102250), but you know more about the runc part than me.

**Comment by kolyshkin** [comment]
So, this PR can be merged then I guess?

**Comment by fromanirh** [comment]
> > So, this PR can be merged then I guess?
> 
> Overall I think it is ok. I just need to think about why the `propagateControllers` was added in the first place. When dealing with systemd (both using systemd driver and procfs driver), systemd will manage controllers and stuff for you. It might be a case where `propagateControllers` had to be run before "update", but in a case where we don't have tests. The whole propagate controllers unless we are in a systemd delegated subtree is quite dodgy. I'll defer that to @giuseppe.
> 
> We don't have any testing of using cgroupfs on cgroup v2, and I am pretty sure running that while using systemd _will_ break, no matter what... :/
> 
> /cc @giuseppe

my 2c: I had a review as well and I fully agree with @odinuge , from my POV all the changes in this PR make sense, with the caveat above about `propagateControllers`. From my (admittedly quick) review of the history, I'm not sure this was not just a accidental duplicate of runc features, but better indeed wait for review from. @giuseppe

File: pkg/kubelet/cm/cgroup_manager_linux.go
Comment (fromanirh) [reply pkg/kubelet/cm/cgroup_manager_linux.go:606]: /cc @sjenning 
at least according to github UI it seems he was not assigned (just mentioned)

**Comment by giuseppe** [comment]
> ```
> 
> ```
> 
> 
>     
>   
> 
> As far as I understand, there are two issues here.
> 
> 
>     1. The test expects `MemoryStats.AvailableBytes` to be `nil` for system containers, and yet it reports something close to 4GB for `runtime` and `kubelet` (but not for `pods`).
> 
> 
>     2. Higher than expected number of `MemoryStats.MajorPageFaults` from the `pods` system container.
> 
> 
> 
> 
> This is _presumably_ caused by the first commit (removing `propagateControllers`), and I can't think about why is that. Perhaps somehow writing to `cgroup.subtree_control` reset the state of the controller?

I forgot what tests was failing but there are two fixes in cAdvisor that AFAICS are not yet propagated into a release and into Kubernetes:

https://github.com/google/cadvisor/pull/2837
https://github.com/google/cadvisor/pull/2839

Especially the first one, solves an issue when reading memory stats

**Comment by giuseppe** [comment]
> Overall I think it is ok. I just need to think about why the `propagateControllers` was added in the first place. When dealing with systemd (both using systemd driver and procfs driver), systemd will manage controllers and stuff for you. It might be a case where `propagateControllers` had to be run before "update", but in a case where we don't have tests. The whole propagate controllers unless we are in a systemd delegated subtree is quite dodgy. I'll defer that to @giuseppe.

at the time I've added that code, runc didn't work with cgroupfs on cgroup v2.  Now that this functionality is in libcontainer, I agree it is better to remove it

**Comment by kolyshkin** [comment]
> > ```
> > 
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > As far as I understand, there are two issues here.
> > ```
> > 1. The test expects `MemoryStats.AvailableBytes` to be `nil` for system containers, and yet it reports something close to 4GB for `runtime` and `kubelet` (but not for `pods`).
> > 
> > 
> > 2. Higher than expected number of `MemoryStats.MajorPageFaults` from the `pods` system container.
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > This is _presumably_ caused by the first commit (removing `propagateControllers`), and I can't think about why is that. Perhaps somehow writing to `cgroup.subtree_control` reset the state of the controller?
> 
> I forgot what tests was failing but there are two fixes in cAdvisor that AFAICS are not yet propagated into a release and into Kubernetes:
> 
> [google/cadvisor#2837](https://github.com/google/cadvisor/pull/2837)
> [google/cadvisor#2839](https://github.com/google/cadvisor/pull/2839)
> 
> Especially the first one, solves an issue when reading memory stats

Thanks, opened https://github.com/google/cadvisor/issues/2878 and https://github.com/google/cadvisor/pull/2879 (it might need a consent from you to make a google-cla bot happy).

File: pkg/kubelet/cm/types.go
Comment (odinuge) [reply pkg/kubelet/cm/types.go:81]: This still reads `MemoryUsage returns current memory usage of the specified cgroup, as read from the cgroupfs.` tho.

**Review by odinuge — APPROVED** [review]
/lgtm

**Review by giuseppe — APPROVED** [review]
LGTM

**Review by mrunalp — APPROVED** [review]

**Pull request merged**
